### ISSUE TITLE
clarify or rethink 504 "Do not use 'local_action', use 'delegate_to: localhost'"

### ISSUE THREAD

**Issue by NotAFile** [ansible-community/ansible-lint#475]
# Issue Type
- Bug report

# Ansible and Ansible Lint details

```
ansible --version
ansible-lint 4.0.1
ansible-lint --version
ansible 2.7.5
```

Any playbook that uses `local_action` receives a lint of "[504] Do not use 'local_action', use 'delegate_to: localhost'" with a `MEDIUM` severity. However, I can't find any reasons against `local_action`, or hints of deprecation or non-recommendation for `local_action` in the docs. The description of the lint should probably be adjusted to clarify the rationale for warning against it, as it is not clear how one style is better than the other.

**Comment by willthames** [comment]
My main issues with local_action is that it's completely contrary to all ansible module documentation - you have to add a module argument, and it then becomes less readable. 

Having said that, I also prefer `connection: local` to `delegate_to: localhost` but that's an entire other rabbit hole:

https://willthames.github.io/2018/07/01/connection-local-vs-delegate_to-localhost.html

**Comment by awcrosby** [comment]
We can update the rule's `description` to say:

To increase readability and match the style of typical ansible tasks, use ``delegate_to: localhost`` to replicate the functionality of ``local_action``

**Comment by ssbarnea** [comment]
Closing as nobody bothered to propose the rephrasing.

**Comment by codekiln** [comment]
@awcrosby proposed: 
> To increase readability and match the style of typical ansible tasks, use `delegate_to: localhost` to replicate the functionality of `local_action`

My biggest problem with `local_action` is that it's only briefly mentioned [once in a note at the bottom of one page of the ansible docs](https://docs.ansible.com/ansible/latest/user_guide/playbooks_delegation.html#), and the (required?) `module` param is not mentioned.

I'd rephrase it this way: 
> `local_action` is shorthand for `delegate_to: localhost`, but is not as well documented or as frequently used. Use `delegate_to: localhost` to increase consistency and clarity of intent.

**Comment by ssbarnea** [comment]
I am not sure the linter should make statements like "not as well documented or as frequently used" (even if that may be true), but it feels awkward and if I read such a message I would quickly wonder "why?" on "on what base?, "where are the numbers?". 

Frequence of use is a reason for making a recommendation only in the absence of more objective reasons for preferring one to another.

In fact the linter can now have very long description that can include explanations and reasoning. Maybe we should make use use of them.

One reason for which I dislike local_action is that it creates a divergence branch for the syntax, making parsin harder. I do not think there is any special case where local_action can provide something that is not already done by the standard action with delegate, but I may be wrong.

**Comment by ach5948** [comment]
Sorry to necro a very old thread, but I came across this while googling.

One example where local_action provides something delegate_to doesn't is when including tasks that should mostly be run on a separate host but have some tasks that should still be run locally (e.g. configuring a dynamically added host via a mix of on device commands and api calls).

In the example:
`tasks.yaml`
```yaml
- name: Include Tasks
  ansible.builtin.include_tasks:
    file: subtasks.yaml
  args:
    apply:
      delegate_to: host2
```

`subtasks.yaml`
```yaml
- name: Subtask A
  ansible.builtin.command: hostname
  delegate_to: localhost

- name: Subtask B
  local_action: ansible.builtin.command
  args:
    cmd: hostname
```

Subtask A's `delegate_to` will be overridden and it will run on host2 instead of localhost, while Subtask B will still run on localhost.

Also ansible-lint is the only reference I've seen to `local_action` being deprecated. Is there an announcement by ansible somewhere that it's deprecated? It's currently included as an example in the [official ansible documentation](https://docs.ansible.com/projects/ansible/latest/playbook_guide/playbooks_delegation.html#delegating-tasks).
